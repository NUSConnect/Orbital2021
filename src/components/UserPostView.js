import * as firebase from 'firebase'
import React, { useEffect, useState } from 'react'
import {
  Dimensions,
  StyleSheet,
  Text,
  TouchableOpacity,
  View
} from 'react-native'
import {
  Card,
  Divider,
  PostText,
  PostTime,
  UserImg,
  UserInfo,
  UserInfoText,
  UserName
} from '../styles/FeedStyles'
import moment from 'moment'
import { Ionicons } from 'react-native-vector-icons'
import ProgressiveImage from './ProgressiveImage'

const DeviceWidth = Dimensions.get('window').width

const UserPostView = ({ route, item, navigation }) => {
  const userId = item.id.substring(0, 28)
  const [userData, setUserData] = useState(null)
  const [postData, setPostData] = useState(null)

  const getUser = async () => {
    await firebase
      .firestore()
      .collection('users')
      .doc(userId)
      .get()
      .then((documentSnapshot) => {
        if (documentSnapshot.exists) {
          setUserData(documentSnapshot.data())
        }
      })
  }

  const getPost = async () => {
    await firebase.firestore().collection('posts').doc(userId).collection('userPosts').doc(item.id).get()
      .then((documentSnapshot) => {
        setPostData(documentSnapshot.data())
      })
  }

  useEffect(() => {
    getUser()
    getPost()
  }, [])

  return (
    <View>
      <View style={{ alignItems: 'center' }}>
        <Text style={styles.reportId}>
          Post ID: {item.id}
        </Text>
        <View
          style={{
            height: 2,
            width: DeviceWidth,
            backgroundColor: 'rgba(0,0,0,0.5)',
            marginLeft: 10,
            marginRight: 10
          }}
        />
        <Card key={item.id}>
          <View style={styles.container}>
            <View
              style={styles.user}
              testID='user'
            >
              <UserInfo>
                <UserImg
                  source={{
                    uri: userData
                      ? userData.userImg ||
                                      'https://firebasestorage.googleapis.com/v0/b/orbital2021-a4766.appspot.com/o/profile%2Fplaceholder.png?alt=media&token=8050b8f8-493f-4e12-8fe3-6f44bb544460'
                      : 'https://firebasestorage.googleapis.com/v0/b/orbital2021-a4766.appspot.com/o/profile%2Fplaceholder.png?alt=media&token=8050b8f8-493f-4e12-8fe3-6f44bb544460'
                  }}
                />
                <UserInfoText>
                  <UserName>
                    {userData
                      ? userData.name || 'Anonymous User'
                      : 'Anonymous User'}
                  </UserName>
                  <PostTime testID='time'>
                    {postData ? moment(postData.postTime.toDate()).fromNow() : ''}
                  </PostTime>
                </UserInfoText>
              </UserInfo>
            </View>
          </View>
          <PostText testID='post'>{postData ? postData.post : ''}</PostText>
          {postData != null && postData.postImg != null
            ? (
              <ProgressiveImage
                defaultImageSource={require('../assets/default-img.jpg')}
                source={{ uri: postData.postImg }}
                style={{ width: '100%', height: 350 }}
                resizeMode='contain'
                testID='image'
              />
              )
            : (
              <Divider />
              )}

          <View style={styles.interactionWrapper}>
            <View style={styles.interactionWrapper}>
              <Ionicons
                name='heart-outline'
                size={25}
                color='#333'
              />
              <Text style={{ fontSize: 16 }}>
                {postData === null || postData.likeCount === 0
                  ? '0 Likes'
                  : postData.likeCount === 1
                    ? '1 Like'
                    : postData.likeCount + ' Likes'}
              </Text>
            </View>
          </View>
        </Card>
        <View style={styles.buttonContainer}>
          <TouchableOpacity
            style={styles.button}
            onPress={() => navigation.navigate('CommentScreen', { item: postData })}
          >
            <Text style={styles.buttonText}>
              View Full Post
            </Text>
          </TouchableOpacity>
        </View>
        <View style={styles.buttonContainer}>
          <TouchableOpacity
            style={styles.button}
            onPress={() => navigation.navigate('UnrestrictedViewProfileScreen', { itemId: userId })}
          >
            <Text style={styles.buttonText}>
              View Profile
            </Text>
          </TouchableOpacity>
        </View>
        <View
          style={{
            height: 2,
            width: DeviceWidth,
            backgroundColor: 'rgba(0,0,0,0.5)',
            marginLeft: 10,
            marginRight: 10
          }}
        />
      </View>
      <Text style={styles.subHeader}>
        Reporters:
      </Text>
    </View>
  )
}

export default UserPostView

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    width: DeviceWidth,
    alignItems: 'center'
  },
  user: {
    width: 360
  },
  interactionWrapper: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 4
  },
  buttonContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    height: 50,
    marginVertical: 10
  },
  button: {
    height: 50,
    width: '50%',
    backgroundColor: 'darkorange',
    borderRadius: 30,
    alignItems: 'center',
    justifyContent: 'center'
  },
  buttonText: {
    fontSize: 18,
    color: 'white'
  },
  reportId: {
    fontSize: 18,
    color: 'crimson',
    paddingLeft: 10,
    margin: 6,
    flexWrap: 'wrap',
    width: DeviceWidth
  },
  subHeader: {
    fontSize: 20,
    paddingLeft: 20,
    marginTop: 10,
    marginBottom: 10
  }
})
