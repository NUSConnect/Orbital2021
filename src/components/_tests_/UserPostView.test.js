import * as React from 'react'
import UserPostView from '../UserPostView'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

jest.useFakeTimers()
jest.mock('react-native-vector-icons')

// Mock
jest.mock('firebase', () => {
  return {
    auth: jest.fn().mockReturnValue({ currentUser: { uid: 22222 } })
  }
})

const mockItem = {
  id: 11111,
  userId: 22222
}

// Tests

it('renders correctly', () => {
  const tree = renderer.create(<UserPostView item={mockItem} />)
  expect(tree).toMatchSnapshot()
})
