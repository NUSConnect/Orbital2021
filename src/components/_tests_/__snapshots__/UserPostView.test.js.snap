// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "crimson",
          "flexWrap": "wrap",
          "fontSize": 18,
          "margin": 6,
          "paddingLeft": 10,
          "width": 750,
        }
      }
    >
      Post ID: 
      11111
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.5)",
          "height": 2,
          "marginLeft": 10,
          "marginRight": 10,
          "width": 750,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#f8f8f8",
            "borderRadius": 10,
            "marginBottom": 20,
            "width": "100%",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "width": 750,
          }
        }
      >
        <View
          style={
            Object {
              "width": 360,
            }
          }
          testID="user"
        >
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                  "paddingBottom": 15,
                  "paddingLeft": 15,
                  "paddingRight": 15,
                  "paddingTop": 15,
                },
              ]
            }
          >
            <Image
              source={
                Object {
                  "uri": "https://firebasestorage.googleapis.com/v0/b/orbital2021-a4766.appspot.com/o/profile%2Fplaceholder.png?alt=media&token=8050b8f8-493f-4e12-8fe3-6f44bb544460",
                }
              }
              style={
                Array [
                  Object {
                    "borderRadius": 25,
                    "height": 50,
                    "width": 50,
                  },
                ]
              }
            />
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "column",
                    "justifyContent": "center",
                    "marginLeft": 10,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontSize": 14,
                      "fontWeight": "bold",
                    },
                  ]
                }
              >
                Anonymous User
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#666",
                      "fontSize": 12,
                    },
                  ]
                }
                testID="time"
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
      <Text
        style={
          Array [
            Object {
              "fontSize": 14,
              "marginBottom": 15,
              "paddingLeft": 15,
              "paddingRight": 15,
            },
          ]
        }
        testID="post"
      >
        ---Deleted Post---
      </Text>
      <View
        style={
          Array [
            Object {
              "alignSelf": "center",
              "borderBottomColor": "#dddddd",
              "borderBottomWidth": 1,
              "marginTop": 15,
              "width": "92%",
            },
          ]
        }
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 4,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 4,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 16,
              }
            }
          >
            0 Likes
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "marginVertical": 10,
          "width": "100%",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "darkorange",
            "borderRadius": 30,
            "height": 50,
            "justifyContent": "center",
            "opacity": 1,
            "width": "50%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
            }
          }
        >
          View Full Post
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "marginVertical": 10,
          "width": "100%",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "darkorange",
            "borderRadius": 30,
            "height": 50,
            "justifyContent": "center",
            "opacity": 1,
            "width": "50%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
            }
          }
        >
          View Profile
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.5)",
          "height": 2,
          "marginLeft": 10,
          "marginRight": 10,
          "width": 750,
        }
      }
    />
  </View>
  <Text
    style={
      Object {
        "fontSize": 20,
        "marginBottom": 10,
        "marginTop": 10,
        "paddingLeft": 20,
      }
    }
  >
    Reporters:
  </Text>
</View>
`;
